pages/api/arweave/upload.ts

```diff
@@ -0,0 +1,138 @@
+import { ArweaveFile } from '@/modules/arweave/types';
+import { ArweaveUploadPayload, FILE_FORM_NAME, PAYLOAD_FORM_NAME } from '@/modules/arweave/upload';
+import { parseNotarized, unpackNotarized, verifyNaclSelfContained } from '@/modules/notary';
+import { ApiError, FormData } from '@/modules/utils';
+import { WALLETS } from '@/modules/wallet/server';
+import { PublicKey } from '@solana/web3.js';
+import Ajv, { JTDSchemaType } from 'ajv/dist/jtd';
+import { sha256 } from 'crypto-hash';
+import { File, IncomingForm } from 'formidable';
+import fs from 'fs/promises';
+import { NextApiRequest, NextApiResponse } from 'next';
+
+/** JSON schemas for parsing request parameters. */
+const SCHEMAS = (() => {
+  const ajv = new Ajv();
+
+  const payload: JTDSchemaType<ArweaveUploadPayload> = {
+    properties: {
+      pubkey: { type: 'string' },
+      fileHash: { type: 'string' },
+    },
+  };
+
+  return { parsePayload: ajv.compileParser(payload) };
+})();
+
+/** Verify a notarized post request, returning the file to upload. */
+const verifyPostParams = async (params: FormData) => {
+  const { parsePayload } = SCHEMAS;
+
+  const payloads: string[] | string | undefined = params.fields[PAYLOAD_FORM_NAME];
+  const files: File[] | File | undefined = params.files[FILE_FORM_NAME];
+
+  if (payloads === undefined || files === undefined)
+    throw new ApiError(400, 'Invalid request parameters');
+
+  const payload: string = payloads instanceof Array ? payloads[0] : payloads;
+
+  parsePayload.message = undefined;
+  const payloadDec = await unpackNotarized<ArweaveUploadPayload>(
+    parseNotarized(payload),
+    verifyNaclSelfContained((s) => new PublicKey(s.pubkey).toBuffer()),
+    { parse: parsePayload }
+  );
+
+  if (payloadDec === undefined) {
+    throw new ApiError(
+      400,
+      `Invalid request parameters: ${parsePayload.message ?? 'Signature verification failed'}`
+    );
+  }
+
+  const file: File = files instanceof Array ? files[0] : files;
+
+  const data = await fs.readFile(file.path);
+  const dataArr = new Uint8Array(
+    data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength)
+  );
+
+  const hash = Buffer.from(await sha256(dataArr, { outputFormat: 'buffer' }));
+
+  if (hash.toString('base64') !== payloadDec.fileHash)
+    throw new ApiError(400, 'File hash mismatch');
+
+  if (file.type == null || file.name == null) throw new ApiError(400, 'Invalid file');
+  const { name, type } = file;
+
+  return { data, name, type };
+};
+
+/** Upload a file to Arweave. */
+const postArweaveFile = async ({
+  data,
+  name,
+  type,
+}: {
+  data: Buffer;
+  name: string;
+  type: string;
+}): Promise<ArweaveFile> => {
+  const {
+    arweave,
+    arweaveKeypair: { jwk },
+  } = await WALLETS;
```

Comment:
I had a PR out that moved accessing arweave through next config. I'll look to get that back should help provide consistent access to arweave throught the app and (like it is now only init once).